cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,113 @@
+    { func() };
+};
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CFuncReturnsAndIsCallableWithArgs = requires(TFunc func, TArgs... args) {
+    { func(args...) } -> std::same_as<TReturn>;
+};
+
+template <typename TClass>
+struct FuncInfo : public FuncInfo<decltype(&TClass::operator())> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...)> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...) const> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(*)(TArgs...)> {};
+
+class WorkQueue {
+
+public:
+    WorkQueue() : io_context_(), executor_work_guard_(io_context_.get_executor()) {}
+
+    void run()
+    {
+        // run all scheduled work and wait for executor_work_guard_.reset() to be called from a worker thread
+        io_context_.run();
+
+        // reinitialize io_context_ and executor_work_guard_ to prepare for the next call to run()
+        mutex_.lock();
+        io_context_.restart();
+        new(&executor_work_guard_) boost::asio::executor_work_guard<boost::asio::io_context::executor_type>(io_context_.get_executor());
+        mutex_.unlock();
+    }
+
+    void reset()
+    {
+        // request io_context_.run() to stop executing once all of its scheduled work is finished
+        mutex_.lock();
+        executor_work_guard_.reset();
+        mutex_.unlock();
+    }
+
+    template <typename TFunc>
+    static auto wrapFuncToExecuteInWorkQueueBlocking(WorkQueue& work_queue, TFunc&& func)
+    {
+        return wrapFuncToExecuteInWorkQueueBlockingImpl(work_queue, std::forward<TFunc>(func), FuncInfo<TFunc>());
+    }
+
+private:
+
+    template <typename TFunc, typename TReturn, typename... TArgs> requires CFuncReturnsAndIsCallableWithArgs<TFunc, TReturn, TArgs...>
+    static auto wrapFuncToExecuteInWorkQueueBlockingImpl(WorkQueue& work_queue, TFunc&& func, const FuncInfo<TReturn(*)(TArgs&...)>& fi)
+    {
+        // Note that we capture func by value because we want to guarantee that func is still
+        // accessible after this wrapFuncToExecuteInWorkQueueBlockingImpl(...) function returns.
+
+        // Note also that we assume that the user's function accepts arguments by const reference.
+        // This will avoid an unnecessary copy when the surrounding code (e.g., the RPC server)
+        // calls the function returned here. Deeper inside our implementation, we will eventually
+        // copy these arguments so they are guaranteed to be accessible when we eventually execute
+        // the user's function, but we want to avoid copying wherever possible.
+        return [&work_queue, func](TArgs&... args) -> TReturn {
+
+            // Note that we capture func and args by value because we want to guarantee that they are both
```

Comment:
I agree with your intuition here to inline `scheduleAndExecuteTaskBlocking(...)` here because it's only called once, and doing so would not make the outer function too complicated. But leaving it as a separate function was an intentional choice in my design because we will have multiple of these static helper functions soon, so I think we should leave it as a separate function. I also think it is a bit more readable because multiple layers of lambdas in a single function is confusing.